### DISCUSSION TITLE
Localized blog category names

### DISCUSSION THREAD

**Discussion by dragomano** [squidfunk/mkdocs-material#7047] [Q&A] [answered]
I'm using Mkdocs Material with Blog feature. Locale is 'ru'. How can I define category names to get similar behavior:

Category url: /articles/
Post url: /articles/post-slug/
Category link in the nav: Статьи (/articles)

When I define category via post metadata, I can only setup 'Статьи' (the url will be `/stati`, but I want to get '/articles`), or 'Articles' (but I want to see the 'Статьи' label in the nav). Any ideas? Is there any index file to define category name?

**Comment by squidfunk** [comment]
Thanks for asking. The category slug is generated from the name of the category. You could define a custom [`categories_slugify`](https://squidfunk.github.io/mkdocs-material/plugins/blog/#config.categories_slugify) function that maps categories in Russian to anything else, so you would use `Статьи` in the front matter, and then map that to `articles` or whatever you like.

**Reply by dragomano** [reply]
Could you provide an example of such mapping?

I tried such a code, but it doesn't work:
```yml
      categories_slugify: !!python/object/apply:pymdownx.slugs.slugify
        kwds:
          case: lower
          mapping:
            'Статьи': 'articles'
```

**Reply by alexvoss** [reply]
You would need to write a custom Python function that does the mapping and that replaces the `pymdownx.slugs.slugify` function that is used by default. Note that the code needs to be in your PYTHONPATH, so you might need to set that environment variable when building your project.

**Edit**: something like the following would do the trick. Note that I call the slug-producing function from within the `pymdownx.slugs` module, so get the default behaviour except for the added substitution.

```python
import functools
import pymdownx.slugs

def _make_slug(text, sep, **kwargs):
    slug = pymdownx.slugs._uslugify(text, sep, **kwargs)
    if slug == "hello":
        return "goodbye"
    return slug

def slugify(**kwargs):
    return functools.partial(_make_slug, **kwargs)
```

Put it into a subdirectory `code` or so and make sure it is on the PYTHONPATH, then configure `mkdocs.yml`:

```yaml
plugins:
  - search
  - blog:
      categories_slugify: !!python/object/apply:slugs.slugify
        kwds:
          case: lower
```

**Reply by dragomano** [reply, marked as answer]
Finally, I found a solution to do it with a hook:

```py
def on_pre_page(page, config, files):
    titles = {
        "some_slug/": "Custom Title 1",
        "another_slug/": "Custom Title 2",
    }
    if page.url in titles:
        page.title = titles[page.url]
    return page

def on_page_markdown(markdown, page, config, files):
    replacements = [
        ("# some_slug", "# Custom Title 1"),
        ("# another_slug", "# Custom Title 2"),
    ]
    for old, new in replacements:
        markdown = markdown.replace(old, new)
    return markdown

```

Will this solution have any side effects or is it safe?

**Reply by alexvoss** [reply]
Hooks are there to respond to build events, which is not what is needed here. The code provides a function that Material calls to get a slug. 

Your suggestion made me wonder about the `PYTHONPATH` issue. My initial reaction was to suggest that registering the code as a hook (even though it is not really one) might help to ensure it gets imported. However, that does seem too hacky to me. Better to think of a more principled way to ensure that user code end up on the path.

**Reply by alexvoss** [reply]
I am not sure the hook version you posted solves the original problem? Was the idea not to have one string for the navigation and another for the slug?

**Reply by dragomano** [reply]
I just wanted to get category titles independent of their slugs. For example, by default, a blog category with the 'test' slug has the title 'Test'. What if I want to have 'Another title' without changing the slug? This is all the more relevant for multilingual versions of the blog.

It would be ideal to be able to define the title of blog category via frontmatter in `.meta.yml`, but I don't see how to do that in the free version.

Why not add support for such syntax in the future?
```yml
titles:
  example_slug: My title
  second_slug: My another title
```

**Reply by alexvoss** [reply]
There is only so much that can be made configurable without causing a maintenance and documentation nightmare. That is why there are plenty of ways to customize. 

As you rightly point out, it would be possible to organize the posts into folders according to categories and use the Mate plugin to set both the category and the title. That is assuming that there is only one category per post or that you are happy to add other categories manually and that they don't lead to the same problem.

The Meta plugin is part of the [Chipotle funding goal](https://squidfunk.github.io/mkdocs-material/insiders/#16000-chipotle), so you could help push this feature into the public version.

In the meantime, does the code I posted not do what you need? Unless I misunderstood the original request, it seems that it should.

**Reply by squidfunk** [reply]
> I just wanted to get category titles independent of their slugs [...] This is all the more relevant for multilingual versions of the blog.

This can already be done by using [custom index pages](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#custom-index-pages). You can leave the slug function as-is, and then use a category called `articles`, and create a page `category/articles.md` with a title of `# Статьи`. This will render as follows:

![screenshot-localhost-3000-category-articles-1713262403046](https://github.com/squidfunk/mkdocs-material/assets/932156/46900c7f-523e-4b83-8f9d-adf02c2286d1)

The URL will be `/blog/articles/`. The category title is extracted from the custom index page.

Note that this mandates [custom index page](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#custom-index-pages) support which is currently only available to sponsors. If you have access to [Insiders](https://squidfunk.github.io/mkdocs-material/insiders/), here's a project that demonstrates the idea: [blog-custom-index-pages.zip](https://github.com/squidfunk/mkdocs-material/files/14998499/blog-custom-index-pages.zip)

**Reply by squidfunk** [reply]
Addendum: I came to the conclusion that this can already be done after thinking about this for a while – wasn't aware of this in my original response, in https://github.com/squidfunk/mkdocs-material/discussions/7047#discussioncomment-9119409 – sorry for the inconvenience and troubles 😅
